### PR TITLE
fix kubeadm reset error

### LINKED ISSUE #41001: kubeadm reset error
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):BUG REPORT

**Kubernetes version** (use `kubectl version`): Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.1.395+1df1ad9d34f5dc", GitCommit:"1df1ad9d34f5dce40229ac5d82b1ab94c559d443", GitTreeState:"clean", BuildDate:"2017-02-06T07:11:51Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.1", GitCommit:"424a0e9e827a39c714ee03e1891c43635adabcdc", GitTreeState:"clean", BuildDate:"2017-01-31T02:31:32Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

**What happened**:
```
# ./cluster/kubeadm.sh reset
[preflight] Running pre-flight checks
[reset] Draining node: "ubuntu"
[reset] Failed to cleanup node: [failed to drain node "ubuntu": exit status 1]
[reset] Stopping the kubelet service
[reset] Unmounting mounted directories in "/var/lib/kubelet"
[reset] Removing kubernetes-managed containers
[reset] Deleting contents of stateful directories: [/var/lib/kubelet /etc/cni/net.d /var/lib/etcd]
[reset] Deleting contents of config directories: [/etc/kubernetes/manifests /etc/kubernetes/pki]
[reset] Deleting files: [/etc/kubernetes/admin.conf /etc/kubernetes/kubelet.conf /etc/kubernetes/ca.crt]
```

### PR DESCRIPTION
**Pull request by xilabao**
fix #41001

### CODE REVIEW

File: cmd/kubeadm/app/cmd/reset.go [hunk 1]
```diff
@@ -152,7 +152,7 @@ func drainAndRemoveNode(removeNode bool) error {
 	hostname = strings.ToLower(hostname)
 
 	// TODO: Use the "native" k8s client for this once we're confident the versioned is working
-	kubeConfigPath := filepath.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeconfig.KubeletKubeConfigFileName)
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: Sorry, but this won't work.
`kubeadm reset` should work on every node, and admin.conf only exists on the master.

**Review by luxas — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/cmd/reset.go (hunk 1 above)
Comment (xilabao) [inline cmd/kubeadm/app/cmd/reset.go]: @luxas fixed. add --kubeconfig option. Allow run `sudo kubeadm reset --kubeconfig=./admin.conf` or use default kubeconfig on every node.

**Comment by resouer** [comment]
@k8s-bot ok to test

**Comment by xilabao** [comment]
@luxas PTAL

**Comment by luxas** [comment]
I'm sorry, but this is the wrong approach to solving https://github.com/kubernetes/kubeadm/issues/142.
The `admin.conf` file doesn't exist on nodes, so if we should do this, we should grant `kubelet.conf` to drain itself instead.

**Comment by xilabao** [comment]
```
#./cluster/kubectl.sh drain 127.0.0.1 --kubeconfig=/var/run/kubernetes/kubelet.kubeconfig
Error from server (Forbidden): User "system:node:127.0.0.1" cannot update nodes at the cluster scope. (put nodes 127.0.0.1)
```
Maybe we should need to allow update nodes for system:node? @luxas @liggitt @deads2k PTAL

**Comment by liggitt** [comment]
nodes have the ability to update their status, not their spec.

**Comment by liggitt** [comment]
draining a node is not an operation I would expect a node to run on itself, but one a cluster administrator would run on the node

**Comment by xilabao** [comment]
> draining a node is not an operation I would expect a node to run on itself, but one a cluster administrator would run on the node

@luxas Could we create admin.conf on nodes?

**Comment by liggitt** [comment]
>> draining a node is not an operation I would expect a node to run on itself, but one a cluster administrator would run on the node
>
> @luxas Could we create admin.conf on nodes?

I meant that the  admin performs on the node API object, not literally on the node machine we should not push  cluster admin credentials to the nodes.

**Comment by xilabao** [comment]
@liggitt Thanks for your so detailed explain. Does it mean that `drain node` is not necessary when run `kubeadm reset` in node machine?

**Comment by liggitt** [comment]
after further discussion, the ability for a node to update itself will be added in https://github.com/kubernetes/kubernetes/pull/41345

**Comment by xilabao** [comment]
fixed in #41345, closed

**Pull request closed without merging**